libraries/rush-lib/src/logic/Git.ts

```diff
@@ -368,26 +425,307 @@ export class Git {
   }
 
   public hasUncommittedChanges(): boolean {
-    return this.getUncommittedChanges().length > 0;
+    const gitStatusEntries: Iterable<IGitStatusEntry> = this.getGitStatus();
+    for (const gitStatusEntry of gitStatusEntries) {
+      if (gitStatusEntry.kind !== 'ignored') {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  public hasUnstagedChanges(): boolean {
+    const gitStatusEntries: Iterable<IGitStatusEntry> = this.getGitStatus();
+    for (const gitStatusEntry of gitStatusEntries) {
+      if (
+        gitStatusEntry.kind === 'untracked' ||
+        (gitStatusEntry as IChangedGitStatusEntry).unstagedChangeType !== undefined
+      ) {
+        return true;
+      }
+    }
+
+    return false;
   }
 
   /**
    * The list of files changed but not committed
    */
   public getUncommittedChanges(): ReadonlyArray<string> {
-    const changes: string[] = [];
-    changes.push(...this._getUntrackedChanges());
-    changes.push(...this._getDiffOnHEAD());
+    const result: string[] = [];
+    const gitStatusEntries: Iterable<IGitStatusEntry> = this.getGitStatus();
+    for (const gitStatusEntry of gitStatusEntries) {
+      if (gitStatusEntry.kind !== 'ignored') {
+        result.push(gitStatusEntry.path);
+      }
+    }
 
-    return changes.filter((change) => {
-      return change.trim().length > 0;
-    });
+    return result;
   }
 
   public getTagSeparator(): string {
     return this._rushConfiguration.gitTagSeparator || DEFAULT_GIT_TAG_SEPARATOR;
   }
 
+  public *getGitStatus(): Iterable<IGitStatusEntry> {
+    const gitPath: string = this.getGitPathOrThrow();
+    // See Git.test.ts for example output
+    const output: string = this._executeGitCommandAndCaptureOutput(gitPath, [
+      'status',
+      '--porcelain=2',
```

Comment:
I'd recommend using `--no-renames` to simplify parsing, and expressly pass `--untracked-files=normal` even if it is the default.

Reply:
Also probably `--ignored=no`